lib/elixir/pages/Library Guidelines.md

```diff
@@ -0,0 +1,214 @@
+
+For more information on running your project, see the official [Mix & OTP guide](https://elixir-lang.org/getting-started/mix-otp/introduction-to-mix.html) or [Mix documentation](https://hexdocs.pm/mix/Mix.html).
+
+### Applications with supervision tree
+
+The `mix new` command also allows the `--sup` flag to scaffold an application with a supervision tree out of the box. We talk about supervision trees later on when discussing one of the common anti-patterns when writing libraries.
+
+## Publishing
+
+Writing code is only the first of many steps to publish a package. We strongly recommend developers to:
+
+  * Choose a versioning schema. Elixir requires versions to be in the format `MAJOR.MINOR.PATCH` but the meaning of those numbers is up to you. Most projects choose [Semantic Versioning](https://semver.org/).
+
+  * Write tests. Elixir ships with a test-framework named [ExUnit](https://hexdocs.pm/ex_unit/ExUnit.html) and the project generated by `mix new` includes sample tests and doctests.
+
+  * Write documentation. The Elixir community is proud in treating documentation as a first-class citizen and making documentation easily accessible. Libraries contribute to the status quo by providing complete API documentation with examples for their modules, types and functions. Projects like [ExDoc](https://github.com/elixir-lang/ex_doc) can be used to generate HTML and EPUB documents from the documentation. ExDoc also supports writing "extra pages", like this one that you are reading, that augment the documetnation with tutorials, guides, and references. See the [Writing Documentation](writing-documentation.html) page for more information.
+
+Projects are often made available to other developers [by publishing a Hex package](https://hex.pm/docs/publish). Hex also [supports private packages for organizations](https://hex.pm/pricing).
+
+Documentation can be generated with [ExDoc](https://github.com/elixir-lang/ex_doc) and also made available on [HexDocs](https://hexdocs.pm).
+
+## Anti-patterns
+
+In this section we document common anti-patterns to avoid when writing libraries.
+
+### Avoid using exceptions for control-flow
+
+You should avoid using exceptions for control-flow. For example, instead of:
+
+```elixir
+try do
+  contents = File.read!("some_path_that_may_or_may_not_exist")
+  {:it_worked, contents}
+rescue
+  File.Error ->
+    :it_failed
+end
+```
+
+you should prefer:
+
+```elixir
+case File.read("some_path_that_may_or_may_not_exist") do
+  {:ok, contents} -> {:it_worked, contents}
+  {:error, _} -> :it_failed
+end
+```
+
+As a library author, it is your responsibility to make sure users are not required to use exceptions for control-flow in their applications. You can follow the same convention as Elixir here, using function names without `!` for returning `:ok`/`:error` tuples and appending `!` to those function names for the raising version.
```

Comment:
"raising version" somehow tripped me off when reading - "exception raising version" maybe to be a bit more specific/easier to grasp immediately?

Reply:
"...using a function name without `!` for returning `:ok`/`:error` tuples and appending `!` for a version of the function which raises an exception."